src/MongoDB.Driver.Core/Core/Connections/BinaryConnection.cs

```diff
@@ -587,6 +621,64 @@ public async Task SendMessagesAsync(IEnumerable<RequestMessage> messages, Messag
         }
 
         // private methods
+        private bool AnyMessageNeedsToBeCompressed(IEnumerable<RequestMessage> messages)
+        {
+            return _sendCompressorType.HasValue && messages.Any(m => m.MayBeCompressed);
+        }
+
+        private CompressorType? ChooseSendCompressorTypeIfAny(ConnectionDescription connectionDescription)
+        {
+            var availableCompressors = connectionDescription.AvailableCompressors;
+            return availableCompressors.Count > 0 ? (CompressorType?)availableCompressors[0] : null;
+        }
+
+        private IByteBuffer CompressMessages(
+            IEnumerable<RequestMessage> messages,
+            IByteBuffer uncompressedBuffer,
+            MessageEncoderSettings messageEncoderSettings)
+        {
+            var outputBufferChunkSource = new OutputBufferChunkSource(BsonChunkPool.Default);
+            var compressedBuffer = new MultiChunkBuffer(outputBufferChunkSource);
+
+            using (var uncompressedStream = new ByteBufferStream(uncompressedBuffer, ownsBuffer: false))
+            using (var compressedStream = new ByteBufferStream(compressedBuffer, ownsBuffer: false))
+            {
+                foreach (var message in messages)
+                {
+                    var uncompressedMessageLength = uncompressedStream.ReadInt32();
+                    uncompressedStream.Position -= 4;
+
+                    using (var uncompressedMessageSlice = uncompressedBuffer.GetSlice((int)uncompressedStream.Position, uncompressedMessageLength))
+                    using (var uncompressedMessageStream = new ByteBufferStream(uncompressedMessageSlice, ownsBuffer: false))
+                    {
+                        if (message.MayBeCompressed)
+                        {
+                            CompressMessage(message, uncompressedMessageStream, compressedStream, messageEncoderSettings);
+                        }
+                        else
+                        {
+                            uncompressedMessageStream.EfficientCopyTo(compressedStream);
+                        }
+                    }
+                }
+                compressedBuffer.Length = (int)compressedStream.Length;
```

Comment:
small fix to have an actual buffer size

Reply:
I'm surprised this was necessary, but good catch.